File: Code Generation/responses/python/deepseek/medium/properties-graph.py
from typing import List

class Solution:
    def numberOfComponents(self, properties: List[List[int]], k: int) -> int:
        n = len(properties)
        
        # Convert each property list to a set for efficient intersection
        property_sets = [set(row) for row in properties]
        
        # Build adjacency list
        adj = [[] for _ in range(n)]
        
        for i in range(n):
            for j in range(i + 1, n):
                # Compute intersection size
                intersect_count = len(property_sets[i] & property_sets[j])
                if intersect_count >= k:
                    adj[i].append(j)
                    adj[j].append(i)
        
        # Find connected components using BFS/DFS
        visited = [False] * n
        
        def dfs(node):
            stack = [node]
            visited[node] = True
            while stack:
                curr = stack.pop()
                for neighbor in adj[curr]:
                    if not visited[neighbor]:
                        visited[neighbor] = True
                        stack.append(neighbor)
        
        components = 0
        for i in range(n):
            if not visited[i]:
                components += 1
                dfs(i)
        
        return components

File: Code Generation/responses/cpp/deepseek/medium/properties-graph.cpp
#include <vector>
#include <unordered_set>
#include <queue>
#include <algorithm>

using namespace std;

class Solution {
public:
    int numberOfComponents(vector<vector<int>> properties, int k) {
        int n = properties.size();
        
        // Convert each row to a set for efficient intersection calculation
        vector<unordered_set<int>> sets(n);
        for (int i = 0; i < n; i++) {
            for (int val : properties[i]) {
                sets[i].insert(val);
            }
        }
        
        // Build adjacency list
        vector<vector<int>> adj(n);
        for (int i = 0; i < n; i++) {
            for (int j = i + 1; j < n; j++) {
                int common = 0;
                // Count number of common elements between sets[i] and sets[j]
                for (int val : sets[i]) {
                    if (sets[j].count(val)) {
                        common++;
                        if (common >= k) break;
                    }
                }
                if (common >= k) {
                    adj[i].push_back(j);
                    adj[j].push_back(i);
                }
            }
        }
        
        // Count connected components using BFS/DFS
        vector<bool> visited(n, false);
        int components = 0;
        
        for (int i = 0; i < n; i++) {
            if (!visited[i]) {
                components++;
                // BFS
                queue<int> q;
                q.push(i);
                visited[i] = true;
                
                while (!q.empty()) {
                    int node = q.front();
                    q.pop();
                    
                    for (int neighbor : adj[node]) {
                        if (!visited[neighbor]) {
                            visited[neighbor] = true;
                            q.push(neighbor);
                        }
                    }
                }
            }
        }
        
        return components;
    }
};